mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala

```diff
@@ -211,14 +213,16 @@ private object IDFModel {
     val n = v.size
     v match {
       case SparseVector(size, indices, values) =>
-        val nnz = indices.size
-        val newValues = new Array[Double](nnz)
+        val newElements = new ArrayBuffer[(Int, Double)]
         var k = 0
-        while (k < nnz) {
-          newValues(k) = values(k) * idf(indices(k))
+        while (k < indices.size) {
+          val newValue = values(k) * idf(indices(k))
```

Comment:
Good point about `indices.size`, I've added a commit to move that outside the loop.

The purpose of this ticket and PR is to remove explicit zeros from the output of `IDFModel.transform`. There may be further performance optimizations to be done in this section of code (e.g. the four cases as you suggest). It's not clear to me that the code in this PR is significantly less performant than what was already there, or that the suggested optimized code addresses the original issue. Maybe I'm missing something?